packages/react-vis/src/plot/zoom-handler.js

```diff
@@ -0,0 +1,175 @@
+import React, {useEffect, useState, useCallback, useRef} from 'react';
+import {getCombinedClassName} from '../utils/styling-utils';
+import {getAttributeScale} from '../utils/scales-utils';
+
+const DEFAULT_STATE = {
+  brushing: false,
+  bounds: null,
+  startPosition: null
+};
+
+const useStateWithGet = initialState => {
+  const [state, setState] = useState(initialState);
+  const ref = useRef();
+  ref.current = state;
+  const get = useCallback(() => ref.current, []);
+  return [state, setState, get];
+};
```

Comment:
> With useCallback and useEffect, you won't really have these problems since they would always use the latest values, and their execution is not dependant on the render function.

I am not sure I follow. If I have a `useEffect`declaring `getState` as a dependency, I expect `useEffect` won't be re-run even if the state has changed.

 What do you mean by "execution not dependent on the render function" ?

Reply:
It depends on how the `state` is used within the `useEffect`.  If the lifecycle of the effect doesn't depend on the values of `state`, then there won't be an issue.  
Issues can arise in situations like this though
```
useEffect(() => {
  const state = getState();
  return addEventListener(state.type, () => {...})
}, [getState]);
```
In this case, the lifecycle of the effect is dependant on the value of `state` since the event listener should be created and destroyed based on the value of `state.type`

> What do you mean by "execution not dependent on the render function" ?
This mainly refers to `useCallback`. The majority of times that `useCallback` is used, its memoized function is passed as a callback, ie `onClick`, `onMouseOver`, which are executed via user initiated events.

Reply:
> Issues can arise in situations like this though

Yes I agree, that was the case I meant in my first comment, and why I am not fond of giving a tool like this helper which can bring bugs but that's just my opinion. What K like about hooks is that if you follow the eslint rules and stick to them you should can this kind of bugs. In the meantime, I hear that performance issues can arise with too many re-renders

> This mainly refers to useCallback

Yes again I agree this applies to `useCallback` but I was more on the fence for this statement about `useEffect`

Reply:
> Issues can arise in situations like this though

Yes I agree, that was the case I meant in my first comment, and why I am not fond of giving a tool like this helper which can bring bugs but that's just my opinion. What K like about hooks is that if you follow the eslint rules and stick to them you should can this kind of bugs. In the meantime, I hear that performance issues can arise with too many re-renders

> This mainly refers to useCallback

Yes again I agree this applies to `useCallback` but I was more on the fence for this statement about `useEffect`

Reply:
If this helper brings a lot of value I won't fight over it even if I have concerns

I really like what you suggested in this comment: https://github.com/uber/react-vis/pull/1354#discussion_r438446979

Reply:
Ya.  I misspoke putting `useEffect` in that sentence.  The point that I was trying to make with putting in there was that the value in the state doesn't necessarily affect the lifecycle of the effect.

How do you feel about the explicit approach?
```
const [state, setState] = useState(DEFAULT_STATE);
const stateRef = useRef(state);

const onMouseUp = useCallback(() => {
   const state = stateRef.current;
   ...
}, [])
```

Reply:
That way it is more self-explanatory and as you said we won't be tempted to use it  in places it can cause bugs

Reply:
OK. I'll change it to use the explicit version.  
I'll also make sure to comment it so that the reader knows why the state is being stored in the `ref`